#' Main function. Takes in VCF and GFF data, and outputs a dataframe that records all CDS-altering mutations, identifies their outcomes (non-synonymous, synonymous, frameshift etc.), and predicts the post-mutation translation product. Intergenic mutations will be ignored.
#'
#' @param vcf vcfR object, ideally generated from FreeBayes. Defaults to global environment object 'vcf'.
#' @param gff Reference sequence annotation GFF3 feature table. Ideally generated by Bakta. Defaults to global environment object 'gff'.
#' @param dna_c Reference concatenated DNA sequence, DNAString object. Defaults to global environment object 'dna'.
#' @param dnaContigs Reference fragmented DNA sequence, DNAString object. Defaults to global environment object 'dna_contigs'.
#' @param virulence BETA. Defaults to FALSE. If TRUE, marks all virulence genes, which is indicated by the description of the translation product in the gff file.
#' @param remove_consensus if a mutation should be removed had all the genomes contained that same mutation. Defaults to FALSE and only works when count is TRUE.
#' @param count if the presence absence of the loci should be counted, defaults to FALSE.
#' @param fix_contig_name BETA. Defaults to NA. If the contig names on the GFF file and the VCF file does not match, you can provide a vector of fixed contig names, which matches the names on the gff file. Contig names on the VCF and the dnaContigs will be changed accordingly.
#' 
#' @import  dplyr
#'          vcfR
#'          Biostrings
#'          stringr
#'          ape
#'          tidyr
#'                    
#' @export

vcgff_gen = function(vcf=vcf,gff=gff,dna_c=dna,dnaContigs=dna_contigs,virulence=F,count=F,remove_consensus=F,fix_contig_name=NA){
  
  
  #fix contig names
  ifelse(
    any(!vcf@fix[,1] %in% str_extract(names(dnaContigs),"^[^\\s]*" ) ), stop('vcf name invalid'),
    ifelse( is.na(fix_contig_name) && all(levels(gff$seqid) %in% str_extract(names(dnaContigs),"^[^\\s]*") )  , print("Contig names validated."),
            ifelse(unique(fix_contig_name) %>% length() != length(fix_contig_name) , print("fix_contig_name not unique,fix failed"),
                   ifelse(is.vector(fix_contig_name) && length(fix_contig_name)==length(names(dnaContigs)) && all(levels(gff$seqid) %in% fix_contig_name ), {
                     vcf@fix[,1] <- fix_contig_name[match(as.data.frame(vcf@fix)$CHROM,str_extract(names(dnaContigs),"^[^\\s]*") )];
                     dnaContigs <- dnaContigs %>% `names<-`(fix_contig_name);
                     print("contig names fixed")}, print("fix_contig_name is not a vector of the same length as the number of DNA contigs, fix failed")
                   )
            ))) %>% suppressWarnings()
  
  contig <- get_concat_position(dnaContigs)
  vcf_c <- concat_vcf(vcf,contig,remove_consensus,count)
  gff_c <- concat_gff(gff,contig)
  
  vcf_c1 <- vcf_c%>% 
    rowwise() %>% 
    mutate( "frameshift"= nchar(ALT)-nchar(REF) ) %>% 
    mutate( "type"=case_when(
      frameshift==0 ~ "snp",
      frameshift>0 ~ "ins",
      frameshift<0 ~ "del"
    ) ) %>% 
    mutate( "start"=as.numeric(POS) ) %>% 
    mutate( "end"=start+nchar(ALT) ) %>% 
    rename("seqid"="#CHROM")%>% 
    mutate( "score"=as.numeric(QUAL) ) %>% 
    mutate("strand"="+","source"="vcf")
  
  vcgff <- full_join(gff_c,vcf_c1) %>%
    suppressMessages() %>% 
    arrange(start) %>% 
    rowwise()  %>% 
    mutate( "gene_ID"=str_remove( str_extract( attributes,'ID\\=[^;]+' ), "ID\\=" ) ) %>% 
    mutate( "product"=str_remove( str_extract( attributes,'Name\\=[^;]+' ),"Name\\=" ) ) %>% 
    mutate( "gene_name"=str_remove( str_extract( attributes,'gene\\=[^\"]+' ), "gene\\=" ) ) 
  
  #for loop to relate snps to genes, edit each line of variation
  i=1;n=0;for ( m in which(vcgff$type %in% c("snp","del","ins")) ) {
    
    if(vcgff$seqid[m]!=vcgff$seqid[m-1] || m==1) {
      
      vcgff$strand[m] <- NA
      vcgff$start[m] <- NA
      vcgff$end[m] <- NA
      vcgff$attributes[m] <- "contig_tip"
      vcgff$FILTER[m] <- sprintf( "mut%03d", 1:nrow(vcf_c1) )[i]
      
    }else{
      
      #Checking if multiple variations exist in the same gene/intergenic region
      if( (m-which(vcgff$source == "vcf")[i-1])!=1 || i==1 ) {
        #in case only one variation found in the gene
        #checking if located in gene or intergenic
        if( vcgff$start[m]<=vcgff$end[m-1] ){
          
          vcgff$attributes[m] <- vcgff$attributes[m-1]
          vcgff$start[m] <- vcgff$start[m-1]
          vcgff$end[m] <- vcgff$end[m-1]
          vcgff$gene_ID[m] <- vcgff$gene_ID[m-1]
          vcgff$product[m] <- vcgff$product[m-1]
          vcgff$gene_name[m] <- vcgff$gene_name[m-1]
          vcgff$phase[m]<-vcgff$phase[m-1]
          vcgff$strand[m] <- vcgff$strand[m-1]
          vcgff$FILTER[m] <- sprintf( "mut%03d", 1:nrow(vcf_c1) )[i]
          
        }else{
          if(vcgff$seqid[m]!=vcgff$seqid[m+1] || m==nrow(vcgff) ){
            
            vcgff$strand[m] <- NA
            vcgff$start[m] <- NA
            vcgff$end[m] <- NA
            vcgff$attributes[m] <- "contig_tip"
            vcgff$FILTER[m] <- sprintf( "mut%03d", 1:nrow(vcf_c1) )[i]
            
          }else{
            
            vcgff$strand[m] <- NA
            vcgff$attributes[m] <- paste( "intergenic",vcgff$gene_ID[m-1],vcgff$strand[m-1],sep = "," )
            vcgff$start[m] <- NA
            vcgff$end[m] <- NA
            #to do: find a way to call the next gene row number
            vcgff$FILTER[m] <- sprintf( "mut%03d", 1:nrow(vcf_c1) )[i]
            
          }
        }
        #record the row number of the last gene
        n=m-1
        
      }else{
        if (n==0){
          vcgff$strand[m] <- NA
          vcgff$start[m] <- NA
          vcgff$end[m] <- NA
          vcgff$attributes[m] <- "contig_tip"
          vcgff$FILTER[m] <- sprintf( "mut%03d", 1:nrow(vcf_c1) )[i]
          
        }else{
          #In case multiple variations are found within the same locus,
          #checking if located in gene or intergenic
          if( vcgff$start[m]<vcgff$end[n] ){
            
            vcgff$attributes[m] <- vcgff$attributes[n]
            vcgff$start[m] <- vcgff$start[n]
            vcgff$end[m] <- vcgff$end[n]
            vcgff$gene_ID[m] <- vcgff$gene_ID[n]
            vcgff$product[m] <- vcgff$product[n]
            vcgff$gene_name[m] <- vcgff$gene_name[n]
            vcgff$phase[m]<-vcgff$phase[n]
            vcgff$strand[m] <- vcgff$strand[n]
            vcgff$FILTER[m] <- sprintf( "mut%03d", 1:nrow(vcf_c1) )[i]
            
          }else{
            
            if(vcgff$seqid[m]!=vcgff$seqid[m+1] || m==nrow(vcgff) ){
              
              vcgff$strand[m] <- NA
              vcgff$start[m] <- NA
              vcgff$end[m] <- NA
              vcgff$attributes[m] <- "contig_tip"
              vcgff$FILTER[m] <- sprintf( "mut%03d", 1:nrow(vcf_c1) )[i]
              
            }else{
              
              vcgff$strand[m] <- NA
              vcgff$start[m] <- NA
              vcgff$end[m] <- NA
              vcgff$attributes[m] <- paste("intergenic",vcgff$gene_ID[n],vcgff$strand[n],sep = ",")
              vcgff$FILTER[m] <- sprintf( "mut%03d", 1:nrow(vcf_c1) )[i]
            }
          }
        }
      }
    }
    i=i+1
  }
  
  vcgff<-vcgff %>%
    filter(source == "vcf")%>%
    rowwise() %>%
    mutate(POS=as.numeric(POS)) %>%
    filter( !is.na(strand) ) %>%
    mutate(DNA_length=end-start+1) %>%
    mutate("position_of_mutation"=if( is.na(strand) ){
      NA
    }else{
      if(strand=="-"){
        end-POS+1
      }else{
        POS-start+1
      }
    }
    ) %>%
    mutate( "ref_DNA"= if(strand=="+"){
      substring(dna_c,start,end) %>% as.character()
    }else{
      substring(dna_c,start,end) %>% DNAString() %>% reverseComplement() %>% as.character()
    } ) %>%
    mutate("ref_aa"=translate( DNAString(ref_DNA) ) %>% as.character() ) %>% 
    
    mutate( "mut_DNA"= get_extended_CDS(dna_c,contig,start,end,POS,strand,seqid,REF,ALT)
            #           if(strand=="+"){
            #   paste( substring(dna_c,start,POS-1),ALT,substring(dna_c,POS+nchar(REF),end),sep="" ) %>% as.character()
            # }else{
            #   paste( substring(dna_c,start,POS-1),ALT,substring(dna_c,POS+nchar(REF),end),sep="" ) %>% DNAString() %>% reverseComplement() %>% as.character()
            # }
    ) %>%
    filter(!is.na(mut_DNA) )%>% 
    mutate( "mut_aa"=( translate( DNAString( mut_DNA ) ) %>% suppressWarnings() %>%  as.character() ) ) %>%
    mutate( "alteration"= case_when(
      ( (position_of_mutation<4 && substring(ref_aa,1,1)!=substring(mut_aa,1,1) ) ) ~ paste("start codon mutation (",substring(ref_DNA,1,3),"->",substring(mut_DNA,1,3),")",sep=""),
      ( type=="snp" && ref_aa==mut_aa ) ~ "synonymous substitution",
      ( type=="snp" && nchar(ref_aa)>nchar(mut_aa)) ~ paste("truncation caused by substitution (",nchar(ref_aa)-1,"->",nchar(mut_aa)-1,")",sep=""),
      ( type=="snp" && nchar(ref_aa)<nchar(mut_aa)) ~ paste("extension caused by substitution (",nchar(ref_aa)-1,"->",nchar(mut_aa)-1,")",sep=""),
      (type=="snp" && ref_aa!=mut_aa)  ~ paste("nonsynonymous substitution (",substring(ref_aa,ceiling(position_of_mutation/3),ceiling((position_of_mutation+nchar(REF)-1)/3)),
                                               ceiling(position_of_mutation/3),
                                               substring(mut_aa,ceiling(position_of_mutation/3),ceiling( (position_of_mutation+nchar(ALT)-1)/3)),")",sep=""),
      ( type=="ins" && nchar(mut_aa)-nchar(ref_aa)==frameshift/3 ) ~ if(strand=="+")
      {paste("insertion caused by indel (",substring(ref_aa,position_of_mutation/3,position_of_mutation/3),
             ceiling((position_of_mutation)/3),
             substring(mut_aa,ceiling(position_of_mutation/3),ceiling((position_of_mutation+frameshift)/3)),")",sep="")
      }else{paste("insertion caused by indel (", substring(ref_aa,position_of_mutation/3,position_of_mutation/3),
                  ceiling((position_of_mutation)/3),
                  substring(mut_aa,ceiling((position_of_mutation-frameshift)/3),ceiling(position_of_mutation/3)),")",sep="")
      },
      ( type=="del" && nchar(mut_aa)-nchar(ref_aa)==frameshift/3 ) ~ if(strand=="+"){paste("deletion caused by indel (",substring(ref_aa,ceiling((position_of_mutation)/3),ceiling((position_of_mutation-frameshift)/3)),
                                                                                           ceiling((position_of_mutation)/3),
                                                                                           substring(mut_aa,ceiling(position_of_mutation/3),ceiling(position_of_mutation/3)),")",sep="")
      }else{paste("deletion caused by indel (",substring(ref_aa,ceiling((position_of_mutation+frameshift)/3),ceiling((position_of_mutation)/3)),
                  ceiling((position_of_mutation)/3),
                  substring(mut_aa,ceiling( (position_of_mutation+frameshift)/3),ceiling( (position_of_mutation+frameshift)/3)),")",sep="")
      },
      ( ( (type=="del"||type=="ins") && nchar(ref_aa)>nchar(mut_aa) ) ) ~ paste("truncation caused by indel (",nchar(ref_aa)-1,"->",nchar(mut_aa)-1,")",sep = ""),
      TRUE ~ paste("extension caused by indel (",nchar(ref_aa)-1,"->",nchar(mut_aa)-1,")",sep = "")
    ) ) %>%
    rename("Contig"="seqid") %>%
    rename("ref"="phase") %>% 
    rename("Mutation_id"="FILTER") %>%
    rename("DNA_mutation_position"="position_of_mutation") %>%
    mutate("ref_length_aa"=nchar(ref_aa),"mut_length_aa"=nchar(mut_aa)) %>%
    mutate("factor"=ifelse(str_detect(attributes,"VFDB"),"virulence",ifelse(str_detect(attributes,"resistance"),"resistance",NA) ) ) %>% 
    select(c("Mutation_id","type","POS","REF","ALT","alteration","var_count","frameshift","gene_ID","ref","gene_name","product","factor","start","end","DNA_length","DNA_mutation_position","strand","Contig","ref_length_aa","mut_length_aa","score","attributes","INFO","ref_DNA","mut_DNA","ref_aa","mut_aa"),colnames(gt_count(vcf,count))) %>%
    filter(
      if(virulence==T){
        str_detect(attributes,"VFDB") || str_detect(attributes,"resistance")
      }else{
        TRUE
      }
    )
  return(vcgff)
}

